src/Microsoft.Azure.ServiceBus/Amqp/AmqpMessageConverter.cs

```diff
@@ -34,49 +33,71 @@ static class AmqpMessageConverter
 
         public static AmqpMessage BrokeredMessagesToAmqpMessage(IEnumerable<BrokeredMessage> brokeredMessages, bool batchable)
         {
-            AmqpMessage amqpMessage;
-            if (brokeredMessages.Count() == 1)
-            {
-                BrokeredMessage singleBrokeredMessage = brokeredMessages.Single();
-                amqpMessage = AmqpMessageConverter.ClientGetMessage(singleBrokeredMessage);
-            }
-            else
+            AmqpMessage amqpMessage = null;
+            BrokeredMessage firstBrokeredMessage = null;
+            List<Data> dataList = null;
+            int messageCount = 0;
+            foreach (var brokeredMessage in brokeredMessages)
             {
-                List<Data> dataList = new List<Data>();
-                foreach (BrokeredMessage brokeredMessage in brokeredMessages)
+                if (firstBrokeredMessage == null)
                 {
-                    AmqpMessage amqpMessageItem = AmqpMessageConverter.ClientGetMessage(brokeredMessage);
-                    ArraySegment<byte>[] payload = amqpMessageItem.GetPayload();
-                    BufferListStream buffer = new BufferListStream(payload);
-                    ArraySegment<byte> value = buffer.ReadBytes((int)buffer.Length);
-                    dataList.Add(new Data() { Value = value });
+                    firstBrokeredMessage = brokeredMessage;
                 }
 
-                BrokeredMessage firstBrokeredMessage = brokeredMessages.First();
-
-                amqpMessage = AmqpMessage.Create(dataList);
-                amqpMessage.MessageFormat = AmqpConstants.AmqpBatchedMessageFormat;
-
-                if (firstBrokeredMessage.MessageId != null)
+                if (messageCount == 1)
                 {
-                    amqpMessage.Properties.MessageId = firstBrokeredMessage.MessageId;
-                }
+                    dataList = new List<Data> { ToData(amqpMessage) };
```

Comment:
>dataList [](start = 20, length = 8)

In case of 1 message scenario, we would still do one extra allocation for datalist which is not used here.